build.rs

```diff
@@ -757,8 +762,36 @@ fn configure(interpreter_config: &InterpreterConfig) -> Result<()> {
         _ => {}
     }
 
-    if interpreter_config.shared {
-        println!("cargo:rustc-cfg=Py_SHARED");
+    if env::var_os("CARGO_FEATURE_AUTO_INITIALIZE").is_some() {
+        ensure!(
+            interpreter_config.shared,
+            "The `auto-initialize` feature is enabled, however your python installation only \
+            supports embedding the Python interpreter statically. This is most often seen in \
+            Python distributions compiled from source without the `--enable-shared` configuration \
+            option. For example, this is the default for `pyenv` on macOS.\n\
+            \n\
+            Embedding the Python interpreter statically has many complications and does not yet \
+            have first-class support in PyO3. If you are sure you intend to do this, disable the \
+            `auto-initialize` feature and call `pyo3::prepare_freethreaded_python` directly in \
+            your program.\n\
+            \n\
+            If you are attempting to run tests or a binary which is okay to link dynamically, \
+            install a Python distribution which ships with the Python shared library and direct \
+            PyO3 to use it using a virtual environment or the PYO3_PYTHON environment variable.\n\
```

Comment:
Let's place this `if` sentence first. This is a major case, I guess. Also, I think just noting to use a `--enabled-shared` Python is sufficient. I don't see any reason to refer to virtualenv (can't we also use pyenv?).